controller/instance_manager_controller.go

```diff
@@ -296,140 +295,185 @@ func (imc *InstanceManagerController) syncInstanceManager(key string) (err error
+		return err
+	}
+	// The instance manager pod should be created on the preferred node only.
+	if imc.controllerID != im.Spec.NodeID {
+		return nil
+	}
+
+	// Since `spec.nodeName` is specified during the pod creation,
+	// the node cordon can not prevent the pod being launched.
+	unschedulable, err := imc.ds.IsKubeNodeUnschedulable(im.Spec.NodeID)
+	if err != nil {
+		return err
+	}
+	if unschedulable {
+		return nil
+	}
+
+	if err := imc.createInstanceManagerPod(im); err != nil {
+		if !types.ErrorAlreadyExists(errors.Cause(err)) {
+			return err
 		}
-	} else {
+	}
+	// The instance manager state will be updated in the next reconcile loop.
+
+	return nil
+}
+
+func (imc *InstanceManagerController) syncInstanceManagerAPIVersion(im *longhorn.InstanceManager) error {
+	// Avoid changing API versions when InstanceManagers are state Unknown.
+	// Then once required (in the future), the monitor could still talk with the pod and update processes in some corner cases. e.g., kubelet restart.
+	// But for now this controller will do nothing for Unknown InstanceManagers.
+	if im.Status.CurrentState != longhorn.InstanceManagerStateRunning && im.Status.CurrentState != longhorn.InstanceManagerStateUnknown {
 		im.Status.APIVersion = engineapi.UnknownInstanceManagerAPIVersion
 		im.Status.APIMinVersion = engineapi.UnknownInstanceManagerAPIVersion
+		return nil
 	}
 
-	if err := imc.syncInstanceManagerPDB(im); err != nil {
-		return err
+	if im.Status.CurrentState == longhorn.InstanceManagerStateRunning && im.Status.APIVersion == engineapi.UnknownInstanceManagerAPIVersion {
+		if err := imc.versionUpdater(im); err != nil {
+			return err
+		}
 	}
+	return nil
+}
 
+func (imc *InstanceManagerController) syncMonitor(im *longhorn.InstanceManager) error {
+	enableMonitor := false
+	disableMonitor := false
 	if im.Status.CurrentState == longhorn.InstanceManagerStateRunning {
 		if err := engineapi.CheckInstanceManagerCompatibilty(im.Status.APIMinVersion, im.Status.APIVersion); err != nil {
-			if imc.isMonitoring(im.Name) {
-				log.Infof("Instance manager controller will stop monitoring the incompatible instance manager %v", im.Name)
-				imc.stopMonitoring(im)
-			}
-		} else {
-			if !imc.isMonitoring(im.Name) {
-				switch im.Spec.Type {
-				case longhorn.InstanceManagerTypeEngine:
-					fallthrough
-				case longhorn.InstanceManagerTypeReplica:
-					imc.startMonitoring(im, NewInstanceManagerUpdater(im))
-				default:
-					log.Errorf("BUG: instance manager %v has invalid type %v", im.Name, im.Spec.Type)
-					im.Status.CurrentState = longhorn.InstanceManagerStateError
-				}
-			}
+			disableMonitor = true
+		} else { // The monitor will be enabled only when the InstanceManager is running and compatible.
+			enableMonitor = true
 		}
+	} else if im.Status.CurrentState != longhorn.InstanceManagerStateUnknown { // For now Longhorn won't actively disable or enable monitoring when the InstanceManager is Unknown.
```

Comment:
I assume ignoring the unknown state should be related the state should be transient.

> For some reason the state of the Pod could not be obtained. This phase typically occurs due to an error in communicating with the node where the Pod should be running.

Reply:
Yes. We don't know if the `unknown` is a transient state or not hence there is no need to do anything aggressively.

Once the status of the Ready condition remains `Unknown` or `False` for longer than the `pod-eviction-timeout`, the pod will be marked as terminating then the controller would mark the IM as state `error`.

https://kubernetes.io/docs/concepts/architecture/nodes/#condition